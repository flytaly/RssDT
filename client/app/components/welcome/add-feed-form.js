import React from 'react';
import { Formik } from 'formik';
import styled from 'styled-components';
import * as Yup from 'yup';
import { useMutation } from 'react-apollo-hooks';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { SubmitButton } from '../styled/buttons';
import Input from '../forms/input-with-icon';
import Select from '../forms/select-with-icon';
import periods, { periodNames } from '../../types/digest-periods';
import RssIcon from '../../static/rss-square.svg';
import EmailIcon from '../../static/envelope.svg';
import ClockIcon from '../../static/clock.svg';

const ADD_FEED_MUTATION = gql`
  mutation ADD_FEED_MUTATION(
    $email: String!
    $url: String!
    $period: DigestSchedule
    $locale: String
    $timeZone: String
  ) {
    addFeed(
        email: $email
        feedUrl: $url
        feedSchedule: $period
        locale: $locale
        timeZone: $timeZone
    ) {
        message
    }
  }
`;

// STYLED COMPONENTS
const StyledForm = styled.form`
    align-self: center;
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: center;
    width: ${props => props.theme.cardWidth / 2 - 6}rem;
`;

const FeedTitle = styled.h2`
    align-self: center;
`;

// VALIDATION
const AddFeedSchema = Yup.object().shape({
    url: Yup.string()
        .url('Invalid feed address')
        .required('The field is required'),
    email: Yup.string()
        .email('Invalid email address')
        .required('The field is required'),
});

const AddFeedForm = ({ setMessages, user }) => {
    const addFeed = useMutation(ADD_FEED_MUTATION);
    return (
        <Formik
            initialValues={{ email: user.email, url: 'http://', period: periods.DAILY }}
            validationSchema={AddFeedSchema}
            onSubmit={async (formVariables, { setSubmitting, resetForm }) => {
                const { email, url, period } = formVariables;
                try {
                    const variables = { email, url, period };

                    if (Intl && Intl.DateTimeFormat) {
                        const { timeZone, locale } = Intl.DateTimeFormat().resolvedOptions();
                        variables.timeZone = timeZone;
                        variables.locale = locale;
                    }
                    const { data } = await addFeed({ variables });
                    setMessages({ success: data.addFeed.message });
                    resetForm();
                } catch (error) {
                    setMessages({ error: error.message });
                }
                setSubmitting(false);
            }}
        >
            {({
                values,
                errors,
                touched,
                handleChange,
                handleBlur,
                handleSubmit,
                isSubmitting,
            }) => (
                <StyledForm onSubmit={handleSubmit}>
                    <FeedTitle>Add a feed</FeedTitle>
                    <Input
                        id="url"
                        type="url"
                        IconSVG={RssIcon}
                        touched={touched.url}
                        placeholder="http://..."
                        onChange={handleChange}
                        onBlur={handleBlur}
                        value={values.url}
                        error={errors.url}
                        disabled={isSubmitting}
                        title="The RSS or Atom feed url"
                        required
                    />
                    <Input
                        id="email"
                        type="email"
                        IconSVG={EmailIcon}
                        touched={touched.email}
                        placeholder="Email"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        value={values.email}
                        error={errors.email}
                        disabled={isSubmitting || user.email}
                        title="Email address"
                        required
                    />
                    <Select
                        id="period"
                        IconSVG={ClockIcon}
                        defaultValue={periods.DAILY}
                        disabled={isSubmitting}
                        onChange={handleChange}
                        onBlur={handleBlur}
                        title="Select a digest period"
                    >
                        <option value={periods.EVERYHOUR}>{`${periodNames[periods.EVERYHOUR]} digest`}</option>
                        <option value={periods.EVERY2HOURS}>{`${periodNames[periods.EVERY2HOURS]} digest`}</option>
                        <option value={periods.EVERY3HOURS}>{`${periodNames[periods.EVERY3HOURS]} digest`}</option>
                        <option value={periods.EVERY6HOURS}>{`${periodNames[periods.EVERY6HOURS]} digest`}</option>
                        <option value={periods.EVERY12HOURS}>{`${periodNames[periods.EVERY12HOURS]} digest`}</option>
                        <option value={periods.DAILY}>{`${periodNames[periods.DAILY]} digest`}</option>
                    </Select>
                    <SubmitButton type="submit" disabled={isSubmitting}>
                        {isSubmitting ? 'SUBMITTING...' : 'SUBSCRIBE'}
                    </SubmitButton>
                </StyledForm>)}
        </Formik>);
};

AddFeedForm.propTypes = {
    user: PropTypes.shape({ email: PropTypes.string }),
    setMessages: PropTypes.func.isRequired,
};
AddFeedForm.defaultProps = {
    user: { email: '' },
};

export default AddFeedForm;
export { ADD_FEED_MUTATION };
