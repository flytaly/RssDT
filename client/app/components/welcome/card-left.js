import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { StyledLightHalf as CardHalf } from '../styled/card-half';
import GraphQLError from '../graphql-error';

const Message = styled.div`
    border: 1px solid ${props => props.color};
    color: ${props => props.color};
    padding: 1rem;
    word-break: break-word;
    border-radius: 3px;
`;

const CardLeft = ({ messages: { error, success } }) => (
    <CardHalf>
        <p>Enter address of a feed (rss, atom) and your email to receive updates every chosen period.</p>
        {(error || success) && (
            <Message color={error ? 'red' : 'green'} data-testid="add-feed-message">
                {error && <GraphQLError error={error} />}
                {success && success}
            </Message>)
        }
    </CardHalf>
);
CardLeft.propTypes = {
    messages: PropTypes.shape({
        error: PropTypes.string,
        success: PropTypes.string,
    }),
};
CardLeft.defaultProps = {
    messages: { error: '', success: '' },
};

export default CardLeft;
