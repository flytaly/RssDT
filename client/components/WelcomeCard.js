import React from 'react';
import StyledCard from './styled/Card';
import CardLeft from './CardLeft';
import CardRight from './CardRight';


const WelcomeCard = () => (
    <StyledCard>
        <CardLeft />
        <CardRight />
    </StyledCard>
);

export default WelcomeCard;
