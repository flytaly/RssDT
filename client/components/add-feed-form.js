import React from 'react';
import { Formik } from 'formik';
import styled from 'styled-components';
import * as Yup from 'yup';
import SubmitButton from './styled/submit-button';
import Input from './input-with-icon';
import Select from './select-with-icon';

const Form = styled.form`
    align-self: center;
    display: flex;
    flex-direction: column;
    width: ${props => props.theme.cardWidth / 2 - 6}rem;
`;

const FeedTitle = styled.h2`
    align-self: center;
`;

const AddFeedSchema = Yup.object().shape({
    url: Yup.string()
        .url('Invalid feed address'),
    // .required('Required'),
    email: Yup.string()
        .email('Invalid email address'),
    // .required('Required'),
});

const AddFeedForm = () => (
    <Formik
        initialValues={{ email: '', url: 'http://', period: '24' }}
        validationSchema={AddFeedSchema}
        onSubmit={(values) => {
            console.log(values);
        }}
    >
        {({
            values,
            errors,
            // touched,
            handleChange,
            // handleBlur,
            handleSubmit,
            isSubmitting,
        }) => (
            <Form onSubmit={handleSubmit}>
                <FeedTitle>Add feed</FeedTitle>
                <Input
                    type="url"
                    name="url"
                    placeholder="http://..."
                    onChange={handleChange}
                    value={values.url}
                    error={errors.url}
                    required
                />
                <Input
                    type="email"
                    name="email"
                    placeholder="Email"
                    onChange={handleChange}
                    value={values.email}
                    error={errors.email}
                    required
                />
                <Select name="period" defaultValue="24">
                    <option value="1">Every hour</option>
                    <option value="2">Every 2 hours</option>
                    <option value="3">Every 3 hours</option>
                    <option value="6">Every 6 hours</option>
                    <option value="12">Every 12 hours</option>
                    <option value="24">Daily</option>
                </Select>
                <SubmitButton
                    type="submit"
                    disabled={isSubmitting}
                >
                    {isSubmitting ? 'SUBMITTING...' : 'SUBSCRIBE'}
                </SubmitButton>
            </Form>)
        }
    </Formik>);

export default AddFeedForm;
