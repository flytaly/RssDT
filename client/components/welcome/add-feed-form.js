import React from 'react';
import { Formik } from 'formik';
import styled from 'styled-components';
import * as Yup from 'yup';
import { useMutation } from 'react-apollo-hooks';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { SubmitButton } from '../styled/buttons';
import Input from '../forms/input-with-icon';
import Select from '../forms/select-with-icon';
import periods, { periodNames } from '../../types/digest-periods';
import rssIcon from '../../static/rss-square.svg';
import emailIcon from '../../static/envelope.svg';
import clockIcon from '../../static/clock.svg';

const ADD_FEED_MUTATION = gql`
  mutation ADD_FEED_MUTATION(
    $email: String!
    $url: String!
    $period: DigestSchedule
  ) {
    addFeed(
        email: $email
        feedUrl: $url
        feedSchedule: $period
    ) {
        message
    }
  }
`;

// STYLED COMPONENTS
const StyledForm = styled.form`
    align-self: center;
    display: flex;
    flex-direction: column;
    width: ${props => props.theme.cardWidth / 2 - 6}rem;
`;

const FeedTitle = styled.h2`
    align-self: center;
`;

// VALIDATION
const AddFeedSchema = Yup.object().shape({
    url: Yup.string()
        .url('Invalid feed address')
        .required('The field is required'),
    email: Yup.string()
        .email('Invalid email address')
        .required('The field is required'),
});

const AddFeedForm = ({ setMessages }) => {
    const addFeed = useMutation(ADD_FEED_MUTATION);
    return (
        <Formik
            initialValues={{ email: '', url: 'http://', period: periods.DAILY }}
            validationSchema={AddFeedSchema}
            onSubmit={async (variables, { setSubmitting, resetForm }) => {
                const { email, url, period } = variables;
                try {
                    const { data } = await addFeed({ variables: { email, url, period } });
                    setMessages({ success: data.addFeed.message });
                    resetForm();
                } catch (error) {
                    setMessages({ error: error.message });
                }
                setSubmitting(false);
            }}
        >
            {({
                values,
                errors,
                touched,
                handleChange,
                handleBlur,
                handleSubmit,
                isSubmitting,
            }) => (
                <StyledForm onSubmit={handleSubmit}>
                    <FeedTitle>Add a feed</FeedTitle>
                    <Input
                        id="url"
                        type="url"
                        name="url"
                        icon={rssIcon}
                        touched={touched.url}
                        placeholder="http://..."
                        onChange={handleChange}
                        onBlur={handleBlur}
                        value={values.url}
                        error={errors.url}
                        disabled={isSubmitting}
                        title="The RSS or Atom feed url"
                        required
                    />
                    <Input
                        id="email"
                        type="email"
                        name="email"
                        icon={emailIcon}
                        touched={touched.email}
                        placeholder="Email"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        value={values.email}
                        error={errors.email}
                        disabled={isSubmitting}
                        title="Email address"
                        required
                    />
                    <Select
                        id="period"
                        name="period"
                        icon={clockIcon}
                        defaultValue={periods.DAILY}
                        disabled={isSubmitting}
                        onChange={handleChange}
                        onBlur={handleBlur}
                        title="Select a digest period"
                    >
                        <option value={periods.EVERYHOUR}>{`${periodNames[periods.EVERYHOUR]} digest`}</option>
                        <option value={periods.EVERY2HOURS}>{`${periodNames[periods.EVERY2HOURS]} digest`}</option>
                        <option value={periods.EVERY3HOURS}>{`${periodNames[periods.EVERY3HOURS]} digest`}</option>
                        <option value={periods.EVERY6HOURS}>{`${periodNames[periods.EVERY6HOURS]} digest`}</option>
                        <option value={periods.EVERY12HOURS}>{`${periodNames[periods.EVERY12HOURS]} digest`}</option>
                        <option value={periods.DAILY}>{`${periodNames[periods.DAILY]} digest`}</option>
                    </Select>
                    <SubmitButton type="submit" disabled={isSubmitting}>
                        {isSubmitting ? 'SUBMITTING...' : 'SUBSCRIBE'}
                    </SubmitButton>
                </StyledForm>)}
        </Formik>);
};

AddFeedForm.propTypes = {
    setMessages: PropTypes.func.isRequired,
};

export default AddFeedForm;
export { ADD_FEED_MUTATION };
