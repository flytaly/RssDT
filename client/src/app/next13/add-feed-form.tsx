'use client';

import { useState } from 'react';

import ClockIcon from '@/../public/static/clock.svg';
import MailIcon from '@/../public/static/envelope.svg';
import RssSquareIcon from '@/../public/static/rss-square.svg';
import { ValidationError, addFeedAnonAction, addFeedLoggedInAction } from '@/app/_actions';
import { DigestSchedule, periodNames as names } from '@/types';

import InputWithIcon from './icon-input';
import SelectWithIcon from './icon-select';

export function AddFeedForm({ email }: { email?: string }) {
  const [validationError, setValidationError] = useState<ValidationError>(null);
  const [isSubmitting, setIsSubmitting] = useState(false);

  async function processForm(data: FormData) {
    const isLoggedIn = !!email;
    let action = isLoggedIn ? addFeedLoggedInAction : addFeedAnonAction;
    const result = await action(data);
    if (!result) return;

    setValidationError(result.error);
    setIsSubmitting(false);
    if (result.response) {
      console.log(result.response);
    }
  }

  return (
    <form
      className="flex flex-col w-full"
      action={processForm}
      onSubmit={() => setIsSubmitting(true)}
    >
      <InputWithIcon
        id="feed-url"
        name="url"
        IconSVG={RssSquareIcon}
        placeholder="https://..."
        title="The RSS or Atom feed url"
        error={validationError?.url?._errors[0]}
      />
      <InputWithIcon
        id="email"
        name="email"
        IconSVG={MailIcon}
        value={email}
        readOnly={!!email}
        placeholder="Email"
        title="Email address"
        error={validationError?.email?._errors[0]}
      />
      <SelectWithIcon
        id="digest"
        name="digest"
        IconSVG={ClockIcon}
        defaultValue={DigestSchedule.daily}
        error={validationError?.digest?._errors[0]}
        disabled={isSubmitting}
      >
        {Object.values(DigestSchedule).map((sc) => (
          <option key={sc} value={sc}>{`${names[sc]} digest`}</option>
        ))}
      </SelectWithIcon>

      <button
        type="submit"
        className="submit-btn w-full text-xl tracking-wider"
        disabled={isSubmitting}
      >
        {isSubmitting ? 'loading...' : 'subscribe'}
      </button>
    </form>
  );
}
