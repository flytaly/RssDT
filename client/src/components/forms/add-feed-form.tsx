import React from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import Input from './input';
import RssSquareIcon from '../../../public/static/rss-square.svg';
import MailIcon from '../../../public/static/envelope.svg';
import ClockIcon from '../../../public/static/clock.svg';
import Select from './select';
import { periodNames as names, DigestSchedule } from '../../types';
import { useAddFeedWithEmailMutation } from '../../generated/graphql';
import { MessageItem } from '../welcome-card/animated-message';
import GraphQLError from '../graphql-error';

// VALIDATION
const AddFeedSchema = Yup.object().shape({
  url: Yup.string().url('Invalid feed address').required('The field is required'),
  email: Yup.string().email('Invalid email address').required('The field is required'),
});

const getUserInfo = () => {
  if (Intl && Intl.DateTimeFormat) {
    const { timeZone, locale } = Intl.DateTimeFormat().resolvedOptions();
    return { timeZone, locale };
  }
  return null;
};

interface AddFeedFormProps {
  email?: string;
  setMessages?: React.Dispatch<React.SetStateAction<MessageItem[]>>;
}

function makeSuccessMessage(email: string, title: string, schedule: DigestSchedule) {
  const content = (
    <span>
      <div>
        <b>{title}</b>
        <span> [</span>
        <b>{`${names[(schedule as unknown) as DigestSchedule]} digest`}</b>
        <span>] </span>
      </div>
      <div>{`Activation link has been sent to ${email}.`}</div>
    </span>
  );
  return { key: `success${Math.random() * 1000}`, content, type: 'success' } as MessageItem;
}

const AddFeedForm: React.FC<AddFeedFormProps> = ({ email = '', setMessages }) => {
  const [addFeed] = useAddFeedWithEmailMutation();
  console.log(email);

  return (
    <Formik
      initialValues={{ email, url: 'https://', digest: DigestSchedule.daily }}
      enableReinitialize
      validationSchema={AddFeedSchema}
      onSubmit={async (formVariables, { setSubmitting, resetForm }) => {
        try {
          const { data } = await addFeed({
            variables: {
              input: { email: formVariables.email, feedUrl: formVariables.url },
              userInfo: getUserInfo(),
              feedOpts: { schedule: formVariables.digest },
            },
          });
          if (data?.addFeedWithEmail?.userFeed) {
            const { userFeed } = data?.addFeedWithEmail;
            setMessages?.([
              makeSuccessMessage(
                formVariables.email,
                userFeed.feed.title || '',
                (userFeed.schedule as unknown) as DigestSchedule,
              ),
            ]);
            resetForm();
          } else {
            const errMessages: MessageItem[] | undefined = data?.addFeedWithEmail?.errors?.map(
              (e, idx) => ({
                key: `error_${idx + Math.random()}`,
                text: e.message,
                type: 'error',
              }),
            );
            setMessages?.(errMessages || []);
          }
        } catch (err) {
          setMessages?.([{ key: 'error', content: <GraphQLError error={err.message} /> }]);
        }
        setSubmitting(false);
      }}
    >
      {({ values, errors, touched, handleChange, handleBlur, handleSubmit, isSubmitting }) => (
        <form className="flex flex-col w-full" onSubmit={handleSubmit}>
          <Input
            id="url"
            type="url"
            IconSVG={RssSquareIcon}
            placeholder="https://..."
            value={values.url}
            onChange={handleChange}
            onBlur={handleBlur}
            touched={touched.url}
            error={errors.url}
            title="The RSS or Atom feed url"
            disabled={isSubmitting}
            required
          />
          <Input
            id="email"
            type="email"
            IconSVG={MailIcon}
            placeholder="Email"
            value={values.email}
            onChange={handleChange}
            onBlur={handleBlur}
            touched={touched.email}
            error={errors.email}
            disabled={!!email || isSubmitting}
            title="Email address"
            required
          />
          <Select
            id="digest"
            IconSVG={ClockIcon}
            defaultValue={DigestSchedule.daily}
            onChange={handleChange}
            onBlur={handleBlur}
            error={errors.digest}
            disabled={isSubmitting}
          >
            {Object.values(DigestSchedule).map((sc) => (
              <option key={sc} value={sc}>{`${names[sc]} digest`}</option>
            ))}
          </Select>
          <button
            type="submit"
            className="btn w-full text-xl tracking-wider"
            disabled={isSubmitting}
          >
            {isSubmitting ? 'loading...' : 'subscribe'}
          </button>
        </form>
      )}
    </Formik>
  );
};

export default AddFeedForm;
