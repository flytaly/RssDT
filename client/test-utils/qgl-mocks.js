/* eslint-disable import/prefer-default-export */
import ME_QUERY from '../queries/me-query';

export const ME_QUERY_MOCK = {
    request: {
        query: ME_QUERY,
    },
    result: { data: { me: { id: 'id', email: 'email@test.com' } } },
};
