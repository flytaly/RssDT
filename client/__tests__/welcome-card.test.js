import 'jest-dom/extend-expect';
import { render, cleanup, fireEvent, wait, waitForElement, getByText } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import WelcomeCard from '../components/welcome/welcome-card';
import { ADD_FEED_MUTATION } from '../components/welcome/add-feed-form';
import ApolloMockedProvider from '../test-utils/apollo-mocked-provider';
import periods from '../types/digest-periods';
import { ME_QUERY_MOCK, UPDATE_MY_INFO_MOCK } from '../test-utils/qgl-mocks';

expect.extend(toHaveNoViolations);

const values = {
    url: 'http://test.com/feed',
    email: 'email@email.com',
    period: periods.DAILY,
    timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    locale: Intl.DateTimeFormat().resolvedOptions().locale,
};
const successMsg = 'success';
const mocks = [{
    request: {
        query: ADD_FEED_MUTATION,
        variables: {
            ...values,
        },
    },
    result: { data: { addFeed: { message: successMsg } } },
},
ME_QUERY_MOCK,
UPDATE_MY_INFO_MOCK,
];

afterEach(cleanup);

describe('Add feed form', () => {
    test('the form should not have accessibility violations', async () => {
        // It's important to pass container so document.body stays empty
        // because axe temporary appends innerHTML to document.body and hence creates duplicates:
        // https://github.com/nickcolley/jest-axe/issues/56#issuecomment-476872378
        const container = document.createElement('div');
        render(
            <ApolloMockedProvider mocks={mocks}>
                <WelcomeCard />
            </ApolloMockedProvider>,
            { container },
        );

        const result = await axe(container.innerHTML);

        expect(result).toHaveNoViolations();
    });

    test('should show form errors', async () => {
        const { getByLabelText, container } = render(
            <ApolloMockedProvider mocks={mocks}>
                <WelcomeCard />
            </ApolloMockedProvider>,
        );
        const submitBtn = getByText(container.querySelector('form'), /subscribe/i);
        const inputs = {
            url: getByLabelText(/The RSS or Atom feed url/i),
            email: getByLabelText(/Email/i),
            period: getByLabelText(/Select a digest period/i),
        };
        fireEvent.change(inputs.url, { target: { value: 'notAnEmail' } });
        fireEvent.change(inputs.email, { target: { value: 'http:/notaurl' } });
        fireEvent.blur(inputs.url);
        fireEvent.blur(inputs.email);

        await waitForElement(() => [
            getByText(container, /invalid feed address/i),
            getByText(container, /invalid email address/i),
        ]);
        fireEvent.click(submitBtn);
    });
});

describe('Submitting data', () => {
    test('should submit valid data and display message', async () => {
        const {
            getByLabelText, getByTestId, container,
        } = render(
            <ApolloMockedProvider mocks={mocks}>
                <WelcomeCard />
            </ApolloMockedProvider>,
        );
        const submitBtn = getByText(container.querySelector('form'), /subscribe/i);

        const inputs = {
            url: getByLabelText(/The RSS or Atom feed url/i),
            email: getByLabelText(/Email/i),
            period: getByLabelText(/Select a digest period/i),
        };

        fireEvent.change(inputs.url, { target: { value: values.url } });
        fireEvent.change(inputs.email, { target: { value: values.email } });
        fireEvent.change(inputs.period, { target: { value: values.period } });
        fireEvent.click(submitBtn);

        Object.values(inputs).forEach(input => expect(input).toBeDisabled());
        expect(submitBtn).toBeDisabled();
        expect(submitBtn).toHaveTextContent(/submitting/i);

        await wait(() => {
            Object.values(inputs).forEach(input => expect(input).toBeEnabled());
            expect(submitBtn).toBeEnabled();
            expect(submitBtn).toHaveTextContent(/subscribe/i);

            expect(getByTestId('add-feed-message')).toHaveTextContent(`${successMsg}`);
        });
    });

    test('should submit data and display error', async () => {
        const wrongFeedUrl = 'http://notafeed.com';
        const errorMsg = 'It\'s not a feed';
        const errorMocks = [{
            request: {
                query: ADD_FEED_MUTATION,
                variables: { ...values, url: wrongFeedUrl },
            },
            error: new Error(errorMsg),
        }, ME_QUERY_MOCK, UPDATE_MY_INFO_MOCK];

        const { getByLabelText, getByTestId, container } = render(
            <ApolloMockedProvider mocks={errorMocks}>
                <WelcomeCard />
            </ApolloMockedProvider>,
        );

        const submitBtn = getByText(container.querySelector('form'), /subscribe/i);
        const inputs = {
            url: getByLabelText(/The RSS or Atom feed url/i),
            email: getByLabelText(/Email/i),
            period: getByLabelText(/Select a digest period/i),
        };

        fireEvent.change(inputs.url, { target: { value: wrongFeedUrl } });
        fireEvent.change(inputs.email, { target: { value: values.email } });
        fireEvent.change(inputs.period, { target: { value: values.period } });
        fireEvent.click(submitBtn);

        await wait(() => {
            expect(getByTestId('add-feed-message')).toHaveTextContent(errorMsg);
        });
    });
});

describe('Snapshot', () => {
    test('should match snapshot', () => {
        /* eslint-disable global-require */
        const { ThemeProvider } = require('styled-components');
        const { default: theme } = require('../components/themes/default');

        const { container } = render(
            <ApolloMockedProvider mocks={mocks}>
                <ThemeProvider theme={theme}>
                    <WelcomeCard />
                </ThemeProvider>
            </ApolloMockedProvider>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });
});
