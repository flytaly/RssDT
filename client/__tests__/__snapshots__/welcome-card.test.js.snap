// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #FFFFFF;
  color: #000000;
  width: 72rem;
  min-width: 36rem;
  max-width: calc(100vw - 2rem);
  min-height: 20rem;
  margin: 2rem auto;
  border-radius: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 50%;
  min-width: 36rem;
  max-width: calc(100vw - 2rem);
  padding: 1rem 3rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 50%;
  min-width: 36rem;
  max-width: calc(100vw - 2rem);
  padding: 1rem 3rem;
  background: #f5f5f5;
  border: 1px solid #707070;
  border-radius: 10px;
}

.c10 {
  width: 100%;
  padding: 1rem 0;
  margin-top: 1rem;
  background-color: #ac247d;
  color: #FFFFFF;
  border-radius: 20px;
  font-size: 2rem;
  cursor: pointer;
}

.c10:disabled,
.c10 [disabled] {
  background-color: #ae7f9e;
  cursor: default;
}

.c10:focus,
.c10:hover:not(:disabled) {
  outline: none;
  border: 1px solid #ac247d;
  box-shadow: 0 0 1.5rem #ac247d;
}

.c10:active:not(:disabled) {
  -webkit-transform: scale(0.96);
  -ms-transform: scale(0.96);
  transform: scale(0.96);
}

.c6 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 4.5rem;
  max-width: 100%;
  padding: 1rem;
  margin-bottom: 2.8rem;
  background-color: #e6e6e6;
  border: 1px solid #707070;
  border-radius: 30px;
}

.c6:hover {
  border-width: 2px;
}

.c7 {
  width: 2rem;
  height: 2rem;
}

.c8 {
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  margin-left: 0.5rem;
  height: 100%;
  font-size: 1.6rem;
  border: none;
  outline: none;
  box-shadow: none;
  background-color: #e6e6e6;
}

.c9 {
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  margin-left: 0.5rem;
  height: 100%;
  font-size: 1.6rem;
  border: none;
  outline: none;
  box-shadow: none;
  background: url(data:image/svg+xml;base64,PHN2ZyBpZD0iTGF5ZXJfMSIgZGF0YS1uYW1lPSJMYXllciAxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA0Ljk1IDEwIj48ZGVmcz48c3R5bGU+LmNscy0ye2ZpbGw6IzQ0NDt9PC9zdHlsZT48L2RlZnM+PHRpdGxlPmFycm93czwvdGl0bGU+PHBvbHlnb24gY2xhc3M9ImNscy0yIiBwb2ludHM9IjEuNDEgNC42NyAyLjQ4IDMuMTggMy41NCA0LjY3IDEuNDEgNC42NyIvPjxwb2x5Z29uIGNsYXNzPSJjbHMtMiIgcG9pbnRzPSIzLjU0IDUuMzMgMi40OCA2LjgyIDEuNDEgNS4zMyAzLjU0IDUuMzMiLz48L3N2Zz4=) 100% 50% / 1.6rem no-repeat;
  cursor: pointer;
  background-color: #e6e6e6;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 30rem;
}

.c5 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.3rem;
  list-style: none;
  margin: 0;
  padding: 0;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

.c3 li {
  margin-left: 1rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p>
      Enter address of a feed (rss, atom) and your email to receive updates every chosen period.
    </p>
    
  </div>
  <div
    class="c2"
  >
    <ul
      class="c3"
      data-testid="card-header"
    >
      <li>
        <a
          href="/"
        >
          Add new feed
        </a>
      </li>
      <li>
        <a
          href="/login"
        >
          Log in
        </a>
      </li>
    </ul>
    <form
      class="c4"
    >
      <h2
        class="c5"
      >
        Add a feed
      </h2>
      <label
        class="c6"
        title="The RSS or Atom feed url"
      >
        <img
          alt="url"
          class="c7"
          src="/static/rss-square.svg"
        />
        <input
          aria-label="The RSS or Atom feed url"
          class="c8"
          id="url"
          name="url"
          placeholder="http://..."
          required=""
          title="The RSS or Atom feed url"
          type="url"
          value="http://"
        />
      </label>
      <label
        class="c6"
        title="Email address"
      >
        <img
          alt="email"
          class="c7"
          src="/static/envelope.svg"
        />
        <input
          aria-label="Email address"
          class="c8"
          id="email"
          name="email"
          placeholder="Email"
          required=""
          title="Email address"
          type="email"
          value=""
        />
      </label>
      <label
        class="c6"
        title="Select a digest period"
      >
        <img
          alt="period"
          class="c7"
          src="/static/clock.svg"
        />
        <select
          aria-label="Select a digest period"
          class="c9"
          id="period"
          name="period"
          title="Select a digest period"
        >
          <option
            value="EVERYHOUR"
          >
            1 hour digest
          </option>
          <option
            value="EVERY2HOURS"
          >
            2 hours digest
          </option>
          <option
            value="EVERY3HOURS"
          >
            3 hours digest
          </option>
          <option
            value="EVERY6HOURS"
          >
            6 hours digest
          </option>
          <option
            value="EVERY12HOURS"
          >
            12 hours digest
          </option>
          <option
            selected=""
            value="DAILY"
          >
            Daily digest
          </option>
        </select>
      </label>
      <button
        class="c10"
        type="submit"
      >
        SUBSCRIBE
      </button>
    </form>
  </div>
</div>
`;
